src/isolate.rs

```diff
@@ -181,6 +201,57 @@ impl Isolate {
     unsafe { v8__Isolate__GetNumberOfDataSlots(self) - 1 }
   }
 
+  /// Safe alternative to Isolate::get_data
+  ///
+  /// Warning: will be renamed to get_data_mut() after original unsafe version
+  /// is removed.
+  pub fn get_slot_mut<'a, T: 'static>(&'a self) -> Option<RefMut<'a, T>> {
+    let cell = self.get_annex().slots.get(&TypeId::of::<T>())?;
+    let ref_mut = cell.try_borrow_mut().ok()?;
+    let ref_mut = RefMut::map(ref_mut, |box_any| {
+      let mut_any = &mut **box_any;
+      Any::downcast_mut::<T>(mut_any).unwrap()
+    });
+    Some(ref_mut)
+  }
+
+  /// Safe alternative to Isolate::get_data
+  ///
+  /// Warning: will be renamed to get_data() after original unsafe version is
+  /// removed.
+  pub fn get_slot<'a, T: 'static>(&'a self) -> Option<Ref<'a, T>> {
+    let cell = self.get_annex().slots.get(&TypeId::of::<T>())?;
+    let r = cell.try_borrow().ok()?;
+    Some(Ref::map(r, |box_any| {
+      let a = &**box_any;
+      Any::downcast_ref::<T>(a).unwrap()
+    }))
+  }
+
+  /// Safe alternative to Isolate::set_data
+  ///
+  /// Use with Isolate::get_slot and Isolate::get_slot_mut to associate state
+  /// with an Isolate.
+  ///
+  /// This method gives ownership of value to the Isolate. Exactly one object of
+  /// each type can be associated with an Isolate. If called more than once with
+  /// an object of the same type, the earlier version will be dropped and
+  /// replaced.
```

Comment:
This mechanic is pretty wacky and difficult to describe. I wish we could come up with something that was simpler.